{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'mpi4py'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[1;32m/Users/aaron/Documentos/MCDI/Semestre 2/Computo de alto rendimiento/testy.ipynb Cell 1\u001b[0m line \u001b[0;36m1\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/aaron/Documentos/MCDI/Semestre%202/Computo%20de%20alto%20rendimiento/testy.ipynb#W0sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mmpi4py\u001b[39;00m \u001b[39mimport\u001b[39;00m MPI\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/aaron/Documentos/MCDI/Semestre%202/Computo%20de%20alto%20rendimiento/testy.ipynb#W0sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39msieve_of_eratosthenes\u001b[39m(n, rank, size):\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/aaron/Documentos/MCDI/Semestre%202/Computo%20de%20alto%20rendimiento/testy.ipynb#W0sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m     \u001b[39m# Calculate the chunk size for each process\u001b[39;00m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/aaron/Documentos/MCDI/Semestre%202/Computo%20de%20alto%20rendimiento/testy.ipynb#W0sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m     chunk_size \u001b[39m=\u001b[39m n \u001b[39m/\u001b[39m\u001b[39m/\u001b[39m size\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'mpi4py'"
     ]
    }
   ],
   "source": [
    "from mpi4py import MPI\n",
    "\n",
    "def sieve_of_eratosthenes(n, rank, size):\n",
    "    # Calculate the chunk size for each process\n",
    "    chunk_size = n // size\n",
    "    start = rank * chunk_size + 2  # Start from 2 to n\n",
    "    end = (rank + 1) * chunk_size + 2 if rank != size - 1 else n\n",
    "\n",
    "    # Create a boolean array \"primes[0..n]\" and initialize all entries as true.\n",
    "    # A value in primes[i] will finally be false if i is not a prime, else true.\n",
    "    primes = [True] * (end - start)\n",
    "\n",
    "    p = 2\n",
    "    while p * p <= n:\n",
    "        # If primes[p] is not changed, then it is a prime\n",
    "        if primes[p - start]:\n",
    "            # Update all multiples of p within the process's range\n",
    "            for i in range(max(p * p, (start + p - 1) // p * p), end, p):\n",
    "                primes[i - start] = False\n",
    "        p += 1\n",
    "\n",
    "    return [p for p, is_prime in enumerate(primes, start) if is_prime]\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    comm = MPI.COMM_WORLD\n",
    "    rank = comm.Get_rank()\n",
    "    size = comm.Get_size()\n",
    "\n",
    "    n = 100  # Rango de números para encontrar primos\n",
    "\n",
    "    # Scatter the work among processes\n",
    "    local_primes = sieve_of_eratosthenes(n, rank, size)\n",
    "\n",
    "    # Gather the results from all processes\n",
    "    all_primes = comm.gather(local_primes, root=0)\n",
    "\n",
    "    if rank == 0:\n",
    "        # Combine the lists from all processes to get the final list of primes\n",
    "        prime_numbers = [num for sublist in all_primes for num in sublist]\n",
    "        print(\"Números primos en el rango de 2 a\", n, \":\", prime_numbers)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting mpi4py\n",
      "  Downloading mpi4py-3.1.4.tar.gz (2.5 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m2.5/2.5 MB\u001b[0m \u001b[31m3.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m00:01\u001b[0m0m\n",
      "\u001b[?25h  Installing build dependencies ... \u001b[?25ldone\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25ldone\n",
      "\u001b[?25h  Preparing metadata (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25hBuilding wheels for collected packages: mpi4py\n",
      "  Building wheel for mpi4py (pyproject.toml) ... \u001b[?25lerror\n",
      "  \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \n",
      "  \u001b[31m×\u001b[0m \u001b[32mBuilding wheel for mpi4py \u001b[0m\u001b[1;32m(\u001b[0m\u001b[32mpyproject.toml\u001b[0m\u001b[1;32m)\u001b[0m did not run successfully.\n",
      "  \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m╰─>\u001b[0m \u001b[31m[148 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m running bdist_wheel\n",
      "  \u001b[31m   \u001b[0m running build\n",
      "  \u001b[31m   \u001b[0m running build_src\n",
      "  \u001b[31m   \u001b[0m running build_py\n",
      "  \u001b[31m   \u001b[0m creating build\n",
      "  \u001b[31m   \u001b[0m creating build/lib.macosx-10.9-x86_64-cpython-39\n",
      "  \u001b[31m   \u001b[0m creating build/lib.macosx-10.9-x86_64-cpython-39/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/run.py -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/__init__.py -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/bench.py -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/__main__.py -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py\n",
      "  \u001b[31m   \u001b[0m creating build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/futures\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/futures/_base.py -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/futures\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/futures/server.py -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/futures\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/futures/__init__.py -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/futures\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/futures/_core.py -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/futures\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/futures/pool.py -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/futures\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/futures/aplus.py -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/futures\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/futures/__main__.py -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/futures\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/futures/_lib.py -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/futures\n",
      "  \u001b[31m   \u001b[0m creating build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/util\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/util/pkl5.py -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/util\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/util/dtlib.py -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/util\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/util/__init__.py -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/util\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/py.typed -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/__main__.pyi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/__init__.pyi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/run.pyi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/bench.pyi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/MPI.pyi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/dl.pyi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/__init__.pxd -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/libmpi.pxd -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/MPI.pxd -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py\n",
      "  \u001b[31m   \u001b[0m creating build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/include\n",
      "  \u001b[31m   \u001b[0m creating build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/include/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/include/mpi4py/mpi4py.MPI.h -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/include/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/include/mpi4py/mpi4py.MPI_api.h -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/include/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/include/mpi4py/mpi4py.h -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/include/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/include/mpi4py/mpi4py.i -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/include/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/include/mpi4py/mpi.pxi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/include/mpi4py\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/futures/__main__.pyi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/futures\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/futures/__init__.pyi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/futures\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/futures/_core.pyi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/futures\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/futures/aplus.pyi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/futures\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/futures/server.pyi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/futures\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/futures/pool.pyi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/futures\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/futures/_lib.pyi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/futures\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/util/__init__.pyi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/util\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/util/dtlib.pyi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/util\n",
      "  \u001b[31m   \u001b[0m copying src/mpi4py/util/pkl5.pyi -> build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/util\n",
      "  \u001b[31m   \u001b[0m running build_clib\n",
      "  \u001b[31m   \u001b[0m MPI configuration: [mpi] from 'mpi.cfg'\n",
      "  \u001b[31m   \u001b[0m checking for library 'lmpe' ...\n",
      "  \u001b[31m   \u001b[0m clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -c _configtest.c -o _configtest.o\n",
      "  \u001b[31m   \u001b[0m clang -flat_namespace -undefined suppress _configtest.o -llmpe -o _configtest\n",
      "  \u001b[31m   \u001b[0m ld: library not found for -llmpe\n",
      "  \u001b[31m   \u001b[0m clang: error: linker command failed with exit code 1 (use -v to see invocation)\n",
      "  \u001b[31m   \u001b[0m failure.\n",
      "  \u001b[31m   \u001b[0m removing: _configtest.c _configtest.o\n",
      "  \u001b[31m   \u001b[0m building 'mpe' dylib library\n",
      "  \u001b[31m   \u001b[0m creating build/temp.macosx-10.9-x86_64-cpython-39\n",
      "  \u001b[31m   \u001b[0m creating build/temp.macosx-10.9-x86_64-cpython-39/src\n",
      "  \u001b[31m   \u001b[0m creating build/temp.macosx-10.9-x86_64-cpython-39/src/lib-pmpi\n",
      "  \u001b[31m   \u001b[0m clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -c src/lib-pmpi/mpe.c -o build/temp.macosx-10.9-x86_64-cpython-39/src/lib-pmpi/mpe.o\n",
      "  \u001b[31m   \u001b[0m creating build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/lib-pmpi\n",
      "  \u001b[31m   \u001b[0m clang -shared -undefined dynamic_lookup -Wl,-rpath,/Users/aaron/opt/anaconda3/lib -L/Users/aaron/opt/anaconda3/lib -L/Users/aaron/opt/anaconda3/lib -Wl,-rpath,/Users/aaron/opt/anaconda3/lib -L/Users/aaron/opt/anaconda3/lib -install_name libmpe.dylib build/temp.macosx-10.9-x86_64-cpython-39/src/lib-pmpi/mpe.o -o build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/lib-pmpi/libmpe.dylib\n",
      "  \u001b[31m   \u001b[0m checking for library 'vt-mpi' ...\n",
      "  \u001b[31m   \u001b[0m clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -c _configtest.c -o _configtest.o\n",
      "  \u001b[31m   \u001b[0m clang -flat_namespace -undefined suppress _configtest.o -lvt-mpi -o _configtest\n",
      "  \u001b[31m   \u001b[0m ld: library not found for -lvt-mpi\n",
      "  \u001b[31m   \u001b[0m clang: error: linker command failed with exit code 1 (use -v to see invocation)\n",
      "  \u001b[31m   \u001b[0m failure.\n",
      "  \u001b[31m   \u001b[0m removing: _configtest.c _configtest.o\n",
      "  \u001b[31m   \u001b[0m checking for library 'vt.mpi' ...\n",
      "  \u001b[31m   \u001b[0m clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -c _configtest.c -o _configtest.o\n",
      "  \u001b[31m   \u001b[0m clang -flat_namespace -undefined suppress _configtest.o -lvt.mpi -o _configtest\n",
      "  \u001b[31m   \u001b[0m ld: library not found for -lvt.mpi\n",
      "  \u001b[31m   \u001b[0m clang: error: linker command failed with exit code 1 (use -v to see invocation)\n",
      "  \u001b[31m   \u001b[0m failure.\n",
      "  \u001b[31m   \u001b[0m removing: _configtest.c _configtest.o\n",
      "  \u001b[31m   \u001b[0m building 'vt' dylib library\n",
      "  \u001b[31m   \u001b[0m clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -c src/lib-pmpi/vt.c -o build/temp.macosx-10.9-x86_64-cpython-39/src/lib-pmpi/vt.o\n",
      "  \u001b[31m   \u001b[0m clang -shared -undefined dynamic_lookup -Wl,-rpath,/Users/aaron/opt/anaconda3/lib -L/Users/aaron/opt/anaconda3/lib -L/Users/aaron/opt/anaconda3/lib -Wl,-rpath,/Users/aaron/opt/anaconda3/lib -L/Users/aaron/opt/anaconda3/lib -install_name libvt.dylib build/temp.macosx-10.9-x86_64-cpython-39/src/lib-pmpi/vt.o -o build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/lib-pmpi/libvt.dylib\n",
      "  \u001b[31m   \u001b[0m checking for library 'vt-mpi' ...\n",
      "  \u001b[31m   \u001b[0m clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -c _configtest.c -o _configtest.o\n",
      "  \u001b[31m   \u001b[0m clang -flat_namespace -undefined suppress _configtest.o -lvt-mpi -o _configtest\n",
      "  \u001b[31m   \u001b[0m ld: library not found for -lvt-mpi\n",
      "  \u001b[31m   \u001b[0m clang: error: linker command failed with exit code 1 (use -v to see invocation)\n",
      "  \u001b[31m   \u001b[0m failure.\n",
      "  \u001b[31m   \u001b[0m removing: _configtest.c _configtest.o\n",
      "  \u001b[31m   \u001b[0m checking for library 'vt.mpi' ...\n",
      "  \u001b[31m   \u001b[0m clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -c _configtest.c -o _configtest.o\n",
      "  \u001b[31m   \u001b[0m clang -flat_namespace -undefined suppress _configtest.o -lvt.mpi -o _configtest\n",
      "  \u001b[31m   \u001b[0m ld: library not found for -lvt.mpi\n",
      "  \u001b[31m   \u001b[0m clang: error: linker command failed with exit code 1 (use -v to see invocation)\n",
      "  \u001b[31m   \u001b[0m failure.\n",
      "  \u001b[31m   \u001b[0m removing: _configtest.c _configtest.o\n",
      "  \u001b[31m   \u001b[0m building 'vt-mpi' dylib library\n",
      "  \u001b[31m   \u001b[0m clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -c src/lib-pmpi/vt-mpi.c -o build/temp.macosx-10.9-x86_64-cpython-39/src/lib-pmpi/vt-mpi.o\n",
      "  \u001b[31m   \u001b[0m clang -shared -undefined dynamic_lookup -Wl,-rpath,/Users/aaron/opt/anaconda3/lib -L/Users/aaron/opt/anaconda3/lib -L/Users/aaron/opt/anaconda3/lib -Wl,-rpath,/Users/aaron/opt/anaconda3/lib -L/Users/aaron/opt/anaconda3/lib -install_name libvt-mpi.dylib build/temp.macosx-10.9-x86_64-cpython-39/src/lib-pmpi/vt-mpi.o -o build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/lib-pmpi/libvt-mpi.dylib\n",
      "  \u001b[31m   \u001b[0m checking for library 'vt-hyb' ...\n",
      "  \u001b[31m   \u001b[0m clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -c _configtest.c -o _configtest.o\n",
      "  \u001b[31m   \u001b[0m clang -flat_namespace -undefined suppress _configtest.o -lvt-hyb -o _configtest\n",
      "  \u001b[31m   \u001b[0m ld: library not found for -lvt-hyb\n",
      "  \u001b[31m   \u001b[0m clang: error: linker command failed with exit code 1 (use -v to see invocation)\n",
      "  \u001b[31m   \u001b[0m failure.\n",
      "  \u001b[31m   \u001b[0m removing: _configtest.c _configtest.o\n",
      "  \u001b[31m   \u001b[0m checking for library 'vt.ompi' ...\n",
      "  \u001b[31m   \u001b[0m clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -c _configtest.c -o _configtest.o\n",
      "  \u001b[31m   \u001b[0m clang -flat_namespace -undefined suppress _configtest.o -lvt.ompi -o _configtest\n",
      "  \u001b[31m   \u001b[0m ld: library not found for -lvt.ompi\n",
      "  \u001b[31m   \u001b[0m clang: error: linker command failed with exit code 1 (use -v to see invocation)\n",
      "  \u001b[31m   \u001b[0m failure.\n",
      "  \u001b[31m   \u001b[0m removing: _configtest.c _configtest.o\n",
      "  \u001b[31m   \u001b[0m building 'vt-hyb' dylib library\n",
      "  \u001b[31m   \u001b[0m clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -c src/lib-pmpi/vt-hyb.c -o build/temp.macosx-10.9-x86_64-cpython-39/src/lib-pmpi/vt-hyb.o\n",
      "  \u001b[31m   \u001b[0m clang -shared -undefined dynamic_lookup -Wl,-rpath,/Users/aaron/opt/anaconda3/lib -L/Users/aaron/opt/anaconda3/lib -L/Users/aaron/opt/anaconda3/lib -Wl,-rpath,/Users/aaron/opt/anaconda3/lib -L/Users/aaron/opt/anaconda3/lib -install_name libvt-hyb.dylib build/temp.macosx-10.9-x86_64-cpython-39/src/lib-pmpi/vt-hyb.o -o build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/lib-pmpi/libvt-hyb.dylib\n",
      "  \u001b[31m   \u001b[0m running build_ext\n",
      "  \u001b[31m   \u001b[0m MPI configuration: [mpi] from 'mpi.cfg'\n",
      "  \u001b[31m   \u001b[0m checking for dlopen() availability ...\n",
      "  \u001b[31m   \u001b[0m checking for header 'dlfcn.h' ...\n",
      "  \u001b[31m   \u001b[0m clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include/python3.9 -c _configtest.c -o _configtest.o\n",
      "  \u001b[31m   \u001b[0m success!\n",
      "  \u001b[31m   \u001b[0m removing: _configtest.c _configtest.o\n",
      "  \u001b[31m   \u001b[0m success!\n",
      "  \u001b[31m   \u001b[0m checking for library 'dl' ...\n",
      "  \u001b[31m   \u001b[0m clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include/python3.9 -c _configtest.c -o _configtest.o\n",
      "  \u001b[31m   \u001b[0m clang -flat_namespace -undefined suppress _configtest.o -Lbuild/temp.macosx-10.9-x86_64-cpython-39 -ldl -o _configtest\n",
      "  \u001b[31m   \u001b[0m success!\n",
      "  \u001b[31m   \u001b[0m removing: _configtest.c _configtest.o _configtest\n",
      "  \u001b[31m   \u001b[0m checking for function 'dlopen' ...\n",
      "  \u001b[31m   \u001b[0m clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include/python3.9 -c _configtest.c -o _configtest.o\n",
      "  \u001b[31m   \u001b[0m clang _configtest.o -Lbuild/temp.macosx-10.9-x86_64-cpython-39 -ldl -o _configtest\n",
      "  \u001b[31m   \u001b[0m success!\n",
      "  \u001b[31m   \u001b[0m removing: _configtest.c _configtest.o _configtest\n",
      "  \u001b[31m   \u001b[0m building 'mpi4py.dl' extension\n",
      "  \u001b[31m   \u001b[0m clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -DHAVE_DLFCN_H=1 -DHAVE_DLOPEN=1 -I/Users/aaron/opt/anaconda3/include/python3.9 -c src/dynload.c -o build/temp.macosx-10.9-x86_64-cpython-39/src/dynload.o\n",
      "  \u001b[31m   \u001b[0m clang -bundle -undefined dynamic_lookup -Wl,-rpath,/Users/aaron/opt/anaconda3/lib -L/Users/aaron/opt/anaconda3/lib -L/Users/aaron/opt/anaconda3/lib -Wl,-rpath,/Users/aaron/opt/anaconda3/lib -L/Users/aaron/opt/anaconda3/lib build/temp.macosx-10.9-x86_64-cpython-39/src/dynload.o -Lbuild/temp.macosx-10.9-x86_64-cpython-39 -ldl -o build/lib.macosx-10.9-x86_64-cpython-39/mpi4py/dl.cpython-39-darwin.so\n",
      "  \u001b[31m   \u001b[0m checking for MPI compile and link ...\n",
      "  \u001b[31m   \u001b[0m clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include -fPIC -O2 -isystem /Users/aaron/opt/anaconda3/include -arch x86_64 -I/Users/aaron/opt/anaconda3/include/python3.9 -c _configtest.c -o _configtest.o\n",
      "  \u001b[31m   \u001b[0m _configtest.c:2:10: fatal error: 'mpi.h' file not found\n",
      "  \u001b[31m   \u001b[0m #include <mpi.h>\n",
      "  \u001b[31m   \u001b[0m          ^~~~~~~\n",
      "  \u001b[31m   \u001b[0m 1 error generated.\n",
      "  \u001b[31m   \u001b[0m failure.\n",
      "  \u001b[31m   \u001b[0m removing: _configtest.c _configtest.o\n",
      "  \u001b[31m   \u001b[0m error: Cannot compile MPI programs. Check your configuration!!!\n",
      "  \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \n",
      "  \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[?25h\u001b[31m  ERROR: Failed building wheel for mpi4py\u001b[0m\u001b[31m\n",
      "\u001b[0mFailed to build mpi4py\n",
      "\u001b[31mERROR: Could not build wheels for mpi4py, which is required to install pyproject.toml-based projects\u001b[0m\u001b[31m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "!pip install mpi4py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
